packages/cli/src/ui/components/shared/text-buffer.ts

```diff
@@ -1002,140 +611,502 @@ export function useTextBuffer({
+        lines: [...state.lines],
+        cursorRow: state.cursorRow,
+        cursorCol: state.cursorCol,
+      };
+      return {
+        ...state,
+        ...stateToRestore,
+        redoStack: state.redoStack.slice(0, -1),
+        undoStack: [...state.undoStack, currentSnapshot],
+      };
+    }
+
+    case 'replace_range': {
+      const { startRow, startCol, endRow, endCol, text } = action.payload;
+      if (
+        startRow > endRow ||
+        (startRow === endRow && startCol > endCol) ||
+        startRow < 0 ||
+        startCol < 0 ||
+        endRow >= state.lines.length ||
+        (endRow < state.lines.length && endCol > currentLineLen(endRow))
+      ) {
+        return state; // Invalid range
+      }
+
+      const nextState = pushUndo(state);
+      const newLines = [...nextState.lines];
+
+      const sCol = clamp(startCol, 0, currentLineLen(startRow));
+      const eCol = clamp(endCol, 0, currentLineLen(endRow));
+
+      const prefix = cpSlice(currentLine(startRow), 0, sCol);
+      const suffix = cpSlice(currentLine(endRow), eCol);
+
+      const normalisedReplacement = text
+        .replace(/\r\n/g, '\n')
+        .replace(/\r/g, '\n');
+      const replacementParts = normalisedReplacement.split('\n');
+
+      // Replace the content
+      if (startRow === endRow) {
+        newLines[startRow] = prefix + normalisedReplacement + suffix;
+      } else {
+        const firstLine = prefix + replacementParts[0];
+        if (replacementParts.length === 1) {
+          // Single line of replacement text, but spanning multiple original lines
+          newLines.splice(startRow, endRow - startRow + 1, firstLine + suffix);
+        } else {
+          // Multi-line replacement text
+          const lastLine =
+            replacementParts[replacementParts.length - 1] + suffix;
+          const middleLines = replacementParts.slice(1, -1);
+          newLines.splice(
+            startRow,
+            endRow - startRow + 1,
+            firstLine,
+            ...middleLines,
+            lastLine,
+          );
+        }
+      }
+
+      const finalCursorRow = startRow + replacementParts.length - 1;
+      const finalCursorCol =
+        (replacementParts.length > 1 ? 0 : sCol) +
+        cpLen(replacementParts[replacementParts.length - 1]);
+
+      return {
+        ...nextState,
+        lines: newLines,
+        cursorRow: finalCursorRow,
+        cursorCol: finalCursorCol,
+        preferredCol: null,
+      };
+    }
+
+    case 'move_to_offset': {
+      const { text, offset } = action.payload;
+      const [newRow, newCol] = offsetToLogicalPos(text, offset);
+      return {
+        ...state,
+        cursorRow: newRow,
+        cursorCol: newCol,
+        preferredCol: null,
+      };
+    }
+
+    case 'create_undo_snapshot': {
+      return pushUndo(state);
+    }
+
+    default: {
+      const exhaustiveCheck: never = action;
+      console.warn(`Unknown action encountered: ${exhaustiveCheck}`);
```

Comment:
nit: make an error or throw as this would indicate a significant logic bug probably resulting in the user wondering why the input prompt is behaving badly.